### PR TITLE
Better support for literals in jit script

### PR DESCRIPTION
**Pull request by zou3519**
Addresses #8177

A design doc can be found here: [gist](https://gist.github.com/zou3519/4b7f13f03cc9f3612bd9363e6405fa0a) version or [quip](https://fb.quip.com/azL1AqUckBdo) version

General approach:
- Add NumberType, FloatType, IntType to represent Python numbers, floats and ints.
- Emit these types for python literals
- Change aten_schema such that Scalars are NumberType, int64_t and bool are IntType.
- Emit aten::type_as, prim::NumToTensor, and prim::TensorToNum nodes for tensor-number math. (see examples below)
- Erase NumberType,  prim::NumToTensor, and prim::TensorToNum for ONNX export. Let ONNX export the aten::type_as nodes.

### Tensor/number math
```
import torch
@torch.jit.script
def fn(x):
    return x + 1
```
```
graph(%x : Dynamic) {
  %1 : int = prim::Constant[value={1}]()
  %2 : Dynamic = prim::NumToTensor(%1)
  %3 : Dynamic = aten::type_as(%2, %x)
  %4 : Dynamic = aten::add[alpha={1}](%x, %4)
  return (%5);
}
```

### Number/Number Math
```
import torch
@torch.jit.script
def fn(zero):
    c = 1 + 1
    return zero + c
```
```
graph(%zero : Dynamic) {
  %1 : int = prim::Constant[value={1}]()
  %2 : int = prim::Constant[value={1}]()
  %3 : Dynamic = prim::num_to_tensor(%1)
  %4 : Dynamic = prim::num_to_tensor(%2)
  %5 : Dynamic = aten::add[alpha={1}](%3, %4)
  %c : int = prim::TensorToNum(%6)  # this is the result of the addition
  ...
  return (%13);
}
```

cc @zdevito @jamesr66a 

There are a list of follow up tasks that I will file issues for later. If any of these seem important right now I can fold them into this PR.
- Let's say I called prim::NumToTensor on an int. Should the output be of type Long()? (It is currently DynamicType)
- Allow scalar tensors for Scalar args
- Add bool type to the schema to differentiate between int and bool
  - fix var/std
- Test logical operators and make them follow python semantics
- Support passing in non-tensor arguments to script functions/support returning non-tensor values from script functions
- Change the schema for IntList to List[int] or something similar
- Better support for List[int] or Tuple[int] args
- Add peephole to delete unnecessary type_as. This peephole should be probably be run on specialized graphs (in an unspecialized graph, it is hard to tell if a type_as is unnecessary).
- we should clean up the math handling code. Instead of using if statements on types, we can just register additional builtin ops like __add__ and __le__. Including __rdiv__ style ops to handle the reverse cases. This will make it easier to see how math is getting emitted and ensures the math functions always go through the same type-conversions as built in functions.

### CODE REVIEW

File: torch/csrc/jit/graph_executor.cpp
Comment (zdevito) [inline torch/csrc/jit/graph_executor.cpp]: Is this here because PropagateInputShapes does not understand number types? Can we add a comment if so?

File: torch/csrc/jit/interned_strings.h
Comment (zdevito) [inline torch/csrc/jit/interned_strings.h]: This seems dead.

File: torch/csrc/jit/script/module.h
Comment (zdevito) [inline torch/csrc/jit/script/module.h]: Ok, I don't like how in how many places this is appearing. This also looks like it is because PropagateInputShapes does not work with number types. I think the right fix is to add handling to PropagateInputShapes for the operators that you added. This can be done in a follow up PR, since we are blocked on other things waiting for number type.

File: torch/csrc/jit/test_jit.cpp
Comment (zdevito) [inline torch/csrc/jit/test_jit.cpp]: We should decide _when_ EraseNumberTypes happens. If it is just so the interpreter doesn't have to handle it then the `Code` constructor should be the _only_ one to call it, and the calls in the graph executor should be removed.

File: tools/jit/gen_jit_dispatch.py
Comment (zdevito) [inline tools/jit/gen_jit_dispatch.py]: This should become unnecessary when we remove the implicit conversion from Tensor -> Number, right? We should make a note of it here, because it is quite complicated.

File: torch/csrc/jit/script/compiler.cpp
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: Why did these move into emitSugaredExpr? They are still always returning values, which was the purpose of the emitSimpleExpr node.

**Review by zdevito — APPROVED** [review]
This looks good. The follow up issues that you noted will need to be done, but we can parallelize them across multiple people, so I want to get this merged. The only things that should be fixed are:
* Ensure we are not breaking ONNX tests with this change, in particular, I see an ATen op in one of the export strings.
* Move the math handling code back to emitSimpleExpr unless there is a reason it cannot go there. 

I would also add to the task list: we should clean up the math handling code. Instead of using if statements on types, we can just register additional builtin ops like `__add__` and `__le__`. Including `__rdiv__` style ops to handle the reverse cases. This will make it easier to see how math is getting emitted and ensures the math functions always go through the same type-conversions as built in functions.

File: torch/csrc/jit/graph_executor.cpp
Comment (zou3519) [reply torch/csrc/jit/graph_executor.cpp]: Yes

File: torch/csrc/jit/script/module.h
Comment (zou3519) [reply torch/csrc/jit/script/module.h]: Will add to the list of followup prs

File: torch/csrc/jit/script/compiler.cpp
Comment (zou3519) [reply torch/csrc/jit/script/compiler.cpp]: Good point, I'll move them back into the emitSimpleExpr path

File: test/test_jit.py
Comment (zou3519) [reply test/test_jit.py]: I haven't checked, but I'll make a pass through before this is merged

File: torch/csrc/jit/script/compiler.cpp
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: Oh god, why? This surely deserves an explanation

File: torch/csrc/jit/script/compiler.cpp
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: Use `isTensorSubtype`? It's much more readable than this expression

File: torch/csrc/jit/script/compiler.cpp
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: nit: use `fmap` (does e.g. `reserve` which you forgot here)

File: torch/csrc/jit/script/compiler.cpp
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: Shouldn't we treat POW specially as well?

File: torch/csrc/jit/script/compiler.cpp
Comment (apaszke) [inline torch/csrc/jit/script/compiler.cpp]: Should the UNARY_MINUS stay here? Will this case ever get triggered?

File: test/test_jit.py
Comment (apaszke) [inline test/test_jit.py]: Can you make it an expected failure?

File: test/test_jit.py
Comment (apaszke) [inline test/test_jit.py]: What's blocking us from doing it? The fact that they would be scalar tensors and not Python numbers?

File: test/test_jit.py
Comment (apaszke) [inline test/test_jit.py]: Can we please avoid reusing existing tests, with well-defined scope for checking this extra behavior? If I were to refactor the test in the future, it's possible I would remove this stuff, because it doesn't contribute to `test_if` in any way

File: tools/jit/gen_jit_dispatch.py
Comment (zou3519) [reply tools/jit/gen_jit_dispatch.py]: Most likely yes. There's an edge case that I'm worried about (pytorch allows one to pass in a scalar tensor as both a Scalar and Tensor argument, and if we allow that too in jit script the ordering will matter), but I think Sam is working on getting rid of overloads like this. I'll make a note here.

File: test/test_jit.py
Comment (apaszke) [inline test/test_jit.py]: Is the `0` literal so broken right now that we can't even use it? Why?

File: test/test_jit.py
Comment (apaszke) [inline test/test_jit.py]: Expected failure pls

File: test/test_jit.py
Comment (apaszke) [inline test/test_jit.py]: Why did those start to fail now?

File: test/test_jit.py
Comment (apaszke) [reply test/test_jit.py]: We should probably try to run those tests as expected failures, just to make sure they won't sit there forever for no good reason

File: torch/csrc/jit/script/module.h
Comment (zou3519) [reply torch/csrc/jit/script/module.h]: Oh, this one here in particular is to patch ONNX export, I'll try to see if there's a better place to put it.

File: torch/csrc/jit/script/compiler.cpp
Comment (zou3519) [reply torch/csrc/jit/script/compiler.cpp]: Will change

File: test/test_jit.py
Comment (zou3519) [reply test/test_jit.py]: The reason why I added this in here is that `a` and `b` are both treated as DynamicType. `3 + d` is a IntType. We disallow assigning DynamicType a IntType value, so I had to convert the IntType value to a DynamicType value somewhere.

Comment (apaszke) [reply test/test_jit.py]: Ugh, we should probably relax those constraints. If you can't change tensors to numbers, natural things like
```python
loss = 0.
for o in outputs:
    loss += o.sum()
```
will break.

File: test/test_jit.py
Comment (zou3519) [reply test/test_jit.py]: Yes, the 0 literal is broken for a number of reasons. In this particular case the problem is that `for i in range(100)` assigns `i` a TensorType. In the above test (test_script_for_in_range), we're using `c` as the accumulator and adding `i` to `c`.

If we assign `c = 0` originally, `c` has type IntType. However, `i` has some tensor type, and we disallow reassigning variables to things that have unrelated types. This can be fixed when the `range(...)` expression gives an IntType.

I can add more comments to the code if you think that is fine

File: test/test_jit.py
Comment (apaszke) [reply test/test_jit.py]: Actually can we just fix the broken tests by removing the scalar literals (use tensor math only)? The fact that so many of them needed to be change is a sign that our current strategy is way too strict

File: test/test_jit.py
Comment (zou3519) [reply test/test_jit.py]: I think what's going is the following:

there are two schemas for `var`: (* means that after this point everything is a kwarg).
(1) var(Tensor, int, *, bool, bool, Tensor)
(2) var(Tensor, *, bool)

(1) used to come before (2) in aten_schema.cpp.

I made the following changes to schemas:
- Shorter schemas come before longer schemas
- bool, int are treated as IntType rather than DynamicType like before

So the schemas look like the following in aten_schema:
(2) var(DynamicType, IntType)
(1) var(DynamicType, IntType, IntType, DynamicType)
Now, when you call `torch.var(tensor, dim=1)`, the compiler mistakingly matches it with (2) instead of (1), which is a problem.

I'll put a comment in the code with this (and think about if I can fix this quickly or not)

File: torch/csrc/jit/script/compiler.cpp
Comment (zou3519) [reply torch/csrc/jit/script/compiler.cpp]: Resolved; I moved the other UNARY_MINUS case to this location so that this one is the only one triggered.

File: test/test_jit.py
Comment (zou3519) [reply test/test_jit.py]: Two things:
1) Yes, the fact that they would be scalar tensors and not Python numbers
2) There is a check somewhere that the result of a function is a tensortype. A python number has type NumberType so it gets caught by this check. I haven't gotten around to seeing what happens if I disable that check yet.

File: test/test_jit.py
Comment (zou3519) [reply test/test_jit.py]: Do you mean changing all of the python numbers to scalar tensors, or in general just getting rid of python numbers in the tests and sticking in tensors somehow?

File: test/expect/TestScript.test_onnx_export_script_module_if.expect
Comment (jamesr66a) [reply test/expect/TestScript.test_onnx_export_script_module_if.expect]: Yes, I believe so. The form `expr > 0` seems pretty common and would be expected to work for anyone used to implicit conversions. Can we solve this restricted case by doing a peephole optimization that matches (Constant -> type_as) and cast the Constant node to the expected type? If `expr` is dynamic, we're probably screwed though

Comment (zou3519) [reply test/expect/TestScript.test_onnx_export_script_module_if.expect]: Thanks for the confirmation, @jamesr66a.  For now I think I could eliminate the aten::type_as nodes that I added right before exporting to onnx so that the behavior before this PR is unchanged.

**Comment by zou3519** [comment]
I've addressed the two points that @zdevito brought up:

1) The code doesn't break ONNX export anymore: the EraseNumberTypes pass erases aten::type_as nodes that are added for math operations. I changed it so that EraseNumberTypes is only called in the ONNX export path; shape analysis and interpreter.cpp can handle the casting operations now.
2) Moved the math handling code back to emitSimpleExpr

In addition, I took a look at the tests we skipped. [Here is a list of the ones that are fixed by this PR]( https://gist.github.com/zou3519/fdcaaa86eaad69f8b8a2e6a8d4ab9539) (I've renabled those tests.

File: torch/onnx/utils.py
Comment (zou3519) [inline torch/onnx/utils.py]: @jamesr66a does this look like the right place to put this?

**Comment by zou3519** [comment]
caffe2 test failure looks unrelated

**Review by jamesr66a — APPROVED** [review]
k

**Comment by zou3519** [comment]
Going to merge this PR now to unblock @zdevito and the robustness work being done by @jamesr66a and @wanchaol.

@apaszke I can address your comments about the tests in a followup PR (let me know if you would like me to change the tests to use tensors instead of scalars, or if we should consider allowing assigning tensors to ints, like in the common case you mentioned)

**Comment by apaszke** [comment]
Let's just merge it now, and talk about relaxing the type constraints in the next meeting

**Pull request merged**
